File: Tools/adapters/lib/xlsx.mjs
/**
 * A minimal reader for the one thing statistical offices keep publishing in: `.xlsx`.
 *
 * Four of the national name registers ship a workbook and nothing else -- Spain, Finland,
 * Sweden and the UK all publish an Excel file with no CSV beside it. That is not a format
 * anybody would choose for a build input, but refusing it means refusing the registers, so
 * this reads enough of the format to get the cells out.
 *
 * ## Why this exists rather than a dependency
 *
 * `Tools/adapters` has no package manifest and no `node_modules`, on purpose: every input
 * is a pinned file with an integrity hash, and adding a build-time dependency would put an
 * unpinned tree of somebody else's code inside the process that produces the corpus. A
 * spreadsheet reader is the kind of thing one normally installs, so it is worth saying that
 * this was weighed rather than skipped.
 *
 * ## What it supports, and what it does not
 *
 * An `.xlsx` is a ZIP holding XML. This reads the central directory, inflates the members
 * it needs with `node:zlib`, and pulls values out of `sharedStrings.xml` and the sheet XML.
 *
 * It handles what statistical publications actually contain: shared strings, inline
 * strings, numbers, and blank cells. It does not handle formulas, dates as serial numbers,
 * styles, or anything to do with presentation -- and it does not need to, because a table
 * of names and counts has none of that. A file that needs more will fail loudly on a
 * missing value rather than silently produce a wrong one.
 *
 * Row and column positions come from each cell's own reference (`B7`) rather than from
 * counting, because a sheet omits empty cells entirely: a row with a gap in it would
 * otherwise shift every value after the gap one column to the left.
 */

import { inflateRawSync } from 'node:zlib'

/** End-of-central-directory, central-directory-header, and local-file-header signatures. */
const END_OF_DIRECTORY = 0x06054b50
const DIRECTORY_ENTRY = 0x02014b50

/**
 * Reads the ZIP central directory and returns each member's name and its bytes.
 *
 * Members are found through the directory rather than by scanning for local headers,
 * because a local header may declare its sizes as zero and defer them to a trailing data
 * descriptor -- in which case scanning cannot tell where the member ends.
 */
function unzip(bytes) {
  let end = bytes.length - 22
  while (end >= 0 && bytes.readUInt32LE(end) !== END_OF_DIRECTORY) end -= 1
  if (end < 0) throw new Error('not a zip file — no end-of-central-directory record')

  const count = bytes.readUInt16LE(end + 10)
  let offset = bytes.readUInt32LE(end + 16)
  const members = new Map()

  for (let i = 0; i < count; i++) {
    if (bytes.readUInt32LE(offset) !== DIRECTORY_ENTRY) {
      throw new Error(`zip directory entry ${i} has the wrong signature`)
    }
    const method = bytes.readUInt16LE(offset + 10)
    const compressedSize = bytes.readUInt32LE(offset + 20)
    const nameLength = bytes.readUInt16LE(offset + 28)
    const extraLength = bytes.readUInt16LE(offset + 30)
    const commentLength = bytes.readUInt16LE(offset + 32)
    const localOffset = bytes.readUInt32LE(offset + 42)
    const name = bytes.subarray(offset + 46, offset + 46 + nameLength).toString('utf8')

    // The local header repeats the name and extra field, and its extra field length can
    // differ from the directory's — so the data start has to be computed from the local
    // header rather than assumed.
    const localNameLength = bytes.readUInt16LE(localOffset + 26)
    const localExtraLength = bytes.readUInt16LE(localOffset + 28)
    const dataStart = localOffset + 30 + localNameLength + localExtraLength
    const data = bytes.subarray(dataStart, dataStart + compressedSize)

    if (method === 0) members.set(name, data)
    else if (method === 8) members.set(name, inflateRawSync(data))
    else throw new Error(`${name}: compression method ${method} is not supported`)

    offset += 46 + nameLength + extraLength + commentLength
  }
  return members
}

/** XML entities that appear in spreadsheet text. Order matters: `&amp;` must be last. */
function decodeEntities(text) {
  return text
    .replaceAll('&lt;', '<')
    .replaceAll('&gt;', '>')
    .replaceAll('&quot;', '"')
    .replaceAll('&apos;', "'")
    .replaceAll(/&#(\d+);/g, (_, code) => String.fromCodePoint(Number(code)))
    .replaceAll(/&#x([0-9a-fA-F]+);/g, (_, code) => String.fromCodePoint(parseInt(code, 16)))
    .replaceAll('&amp;', '&')
}

/**
 * The shared string table, which is where most cell text actually lives.
 *
 * A string entry can be split across several `<t>` runs when parts of it are formatted
 * differently, so the runs are joined rather than the first one taken. `Mª DEL CARMEN`
 * arrives in two runs in the Spanish file.
 */
function readSharedStrings(members) {
  const xml = members.get('xl/sharedStrings.xml')
  if (!xml) return []
  const text = xml.toString('utf8')
  const strings = []
  for (const [, entry] of text.matchAll(/<si>([\s\S]*?)<\/si>/g)) {
    let value = ''
    for (const [, run] of entry.matchAll(/<t[^>]*>([\s\S]*?)<\/t>/g)) value += run
    strings.push(decodeEntities(value))
  }
  return strings
}

/** `B7` -> `{ row: 7, column: 1 }`, with the column zero-based. */
function positionOf(reference) {
  const match = /^([A-Z]+)(\d+)$/.exec(reference)
  if (!match) return null
  let column = 0
  for (const character of match[1]) column = column * 26 + (character.charCodeAt(0) - 64)
  return { row: Number(match[2]), column: column - 1 }
}

/** Sheet name to its XML part, resolved through the workbook and its relationships. */
function sheetParts(members) {
  const workbook = members.get('xl/workbook.xml')?.toString('utf8') ?? ''
  const rels = members.get('xl/_rels/workbook.xml.rels')?.toString('utf8') ?? ''

  const targets = new Map()
  for (const [, id, target] of rels.matchAll(/<Relationship Id="([^"]+)"[^>]*Target="([^"]+)"/g)) {
    targets.set(id, target.replace(/^\/?xl\//, '').replace(/^\//, ''))
  }

  const parts = new Map()
  for (const [, attributes] of workbook.matchAll(/<sheet\b([^>]*)>/g)) {
    const name = /name="([^"]*)"/.exec(attributes)?.[1]
    const id = /r:id="([^"]+)"/.exec(attributes)?.[1]
    if (!name || !id) continue
    const target = targets.get(id)
    if (target) parts.set(decodeEntities(name), `xl/${target}`)
  }
  return parts
}

/**
 * Reads a workbook into `{ sheetName: rows }`, each row an array of cell strings.
 *
 * Values come back as strings whatever the cell held, because every caller here is about
 * to parse them itself and a number that arrived as `1063756` is not more trustworthy for
 * having passed through a float.
 */
export function readWorkbook(bytes) {
  const members = unzip(bytes)
  const shared = readSharedStrings(members)
  const parts = sheetParts(members)

  const sheets = {}
  for (const [name, part] of parts) {
    const xml = members.get(part)
    if (!xml) continue
    const text = xml.toString('utf8')
    const rows = []

    for (const [, cells] of text.matchAll(/<row\b[^>]*>([\s\S]*?)<\/row>/g)) {
      let row = []
      let index = null
      for (const [, attributes, body] of cells.matchAll(/<c\b([^>]*)(?:\/>|>([\s\S]*?)<\/c>)/g)) {
        const reference = /r="([A-Z]+\d+)"/.exec(attributes)?.[1]
        const position = reference ? positionOf(reference) : null
        if (position && index === null) index = position.row

        const type = /t="([^"]+)"/.exec(attributes)?.[1]
        let value = ''
        if (body) {
          if (type === 'inlineStr') {
            for (const [, run] of body.matchAll(/<t[^>]*>([\s\S]*?)<\/t>/g)) value += run
            value = decodeEntities(value)
          } else {
            const raw = /<v>([\s\S]*?)<\/v>/.exec(body)?.[1]
            if (raw !== undefined) {
              value = type === 's' ? (shared[Number(raw)] ?? '') : decodeEntities(raw)
            }
          }
        }
        // Placed by its own column rather than pushed, since empty cells are omitted from
        // the XML entirely and pushing would slide everything after a gap leftwards.
        if (position) row[position.column] = value
        else row.push(value)
      }
      row = [...row].map((cell) => cell ?? '')
      rows.push(row)
      void index
    }
    sheets[name] = rows
  }
  return sheets
}

File: Sources/DecoyAdapterKit/XLSX.swift
import Foundation

/// A minimal reader for the one thing statistical offices keep publishing in: `.xlsx`.
///
/// Four national name registers ship a workbook and nothing else — Spain, Finland, Sweden
/// and the UK all publish Excel with no CSV beside it. Refusing the format means refusing
/// the registers.
///
/// ## What changed in the port
///
/// The JavaScript had to parse the ZIP container by hand: Node gives you raw inflate but
/// not zip, so `lib/xlsx.mjs` reads the central directory itself. Swift has neither, but it
/// already requires `unzip` to unpack pinned archives — so the container is handed to the
/// tool that is already a dependency, and about a hundred lines of central-directory
/// parsing simply do not need to exist. A hand-rolled DEFLATE would have been the
/// alternative, which is a great deal of risk for a build step.
///
/// The XML side is a scanner rather than the original's regexes. Same grammar, same
/// results — `XLSXTests` checks it against the real workbooks — but it does not depend on
/// lazy quantifiers behaving identically across two regex engines.
///
/// ## What it supports, and what it does not
///
/// Shared strings, inline strings, numbers and blank cells: what statistical publications
/// actually contain. Not formulas, date serials, or styles. A file needing more fails on a
/// missing value rather than quietly producing a wrong one.
public enum XLSX {

    public enum Failure: Error, CustomStringConvertible {
        case notReadable(path: String, detail: String)
        case missingMember(String, in: String)

        public var description: String {
            switch self {
            case .notReadable(let path, let detail):
                return "could not read \(path) as a workbook: \(detail)"
            case .missingMember(let member, let path):
                return "\(path) has no \(member) — not an xlsx, or one this cannot read"
            }
        }
    }

    // MARK: - Container

    /// A member's bytes, inflated by `unzip`.
    static func member(_ name: String, in workbook: URL) throws -> String? {
        let process = Process()
        process.executableURL = URL(fileURLWithPath: "/usr/bin/env")
        process.arguments = ["unzip", "-p", workbook.path, name]
        let output = Pipe()
        process.standardOutput = output
        process.standardError = Pipe()
        try process.run()
        let data = output.fileHandleForReading.readDataToEndOfFile()
        process.waitUntilExit()
        guard process.terminationStatus == 0, !data.isEmpty else { return nil }
        return String(decoding: data, as: UTF8.self)
    }

    // MARK: - XML scanning

    /// The bodies of every `<tag …>…</tag>`, and the attribute text of each opening tag.
    ///
    /// `selfClosing` is not a nicety, it is the difference between agreeing with the
    /// JavaScript and not. Its cell regex accepts `<c r="B7"/>` — a real cell that happens
    /// to be blank, and dropping it would slide the row — while its *row* regex requires a
    /// closing tag, so a `<row r="3" .../>` spacer row is skipped entirely.
    ///
    /// Treating both the same way inserted one empty row into three of the five real
    /// workbooks, which then shifted every later index and surfaced as two completely
    /// unrelated-looking cell mismatches further down the sheet.
    static func elements(_ tag: String, in xml: String, selfClosing: Bool = true)
        -> [(attributes: String, body: String)]
    {
        var found: [(String, String)] = []
        var index = xml.startIndex
        let open = "<\(tag)"
        let close = "</\(tag)>"

        while let start = xml.range(of: open, range: index..<xml.endIndex) {
            // `<row` must not match `<rowBreak`. The character after the name has to end it.
            let after = start.upperBound
            guard after < xml.endIndex else { break }
            let following = xml[after]
            guard following == " " || following == ">" || following == "/" else {
                index = after
                continue
            }
            guard let tagEnd = xml.range(of: ">", range: after..<xml.endIndex) else { break }
            let attributes = String(xml[after..<tagEnd.lowerBound])

            if attributes.hasSuffix("/") {
                if selfClosing { found.append((String(attributes.dropLast()), "")) }
                index = tagEnd.upperBound
                continue
            }
            guard let bodyEnd = xml.range(of: close, range: tagEnd.upperBound..<xml.endIndex)
            else { break }
            found.append((attributes, String(xml[tagEnd.upperBound..<bodyEnd.lowerBound])))
            index = bodyEnd.upperBound
        }
        return found
    }

    /// `name="value"` out of an opening tag's attribute text.
    static func attribute(_ name: String, in attributes: String) -> String? {
        var index = attributes.startIndex
        while let start = attributes.range(of: "\(name)=\"", range: index..<attributes.endIndex) {
            // Guard against `r:id` matching a search for `id`, and `foo:name` for `name`.
            let before =
                start.lowerBound == attributes.startIndex
                ? " " : attributes[attributes.index(before: start.lowerBound)]
            guard before == " " || before == "\"" else {
                index = start.upperBound
                continue
            }
            guard let end = attributes.range(of: "\"", range: start.upperBound..<attributes.endIndex)
            else { return nil }
            return String(attributes[start.upperBound..<end.lowerBound])
        }
        return nil
    }

    /// The five XML entities, plus numeric references, which appear in real name data.
    static func decodeEntities(_ text: String) -> String {
        guard text.contains("&") else { return text }
        var out = ""
        out.reserveCapacity(text.count)
        var index = text.startIndex
        while index < text.endIndex {
            guard text[index] == "&",
                let semicolon = text.range(of: ";", range: index..<text.endIndex),
                text.distance(from: index, to: semicolon.lowerBound) <= 10
            else {
                out.append(text[index])
                index = text.index(after: index)
                continue
            }
            let entity = String(text[text.index(after: index)..<semicolon.lowerBound])
            switch entity {
            case "amp": out.append("&")
            case "lt": out.append("<")
            case "gt": out.append(">")
            case "quot": out.append("\"")
            case "apos": out.append("'")
            default:
                if entity.hasPrefix("#x") || entity.hasPrefix("#X"),
                    let value = UInt32(entity.dropFirst(2), radix: 16),
                    let scalar = Unicode.Scalar(value)
                {
                    out.unicodeScalars.append(scalar)
                } else if entity.hasPrefix("#"), let value = UInt32(entity.dropFirst()),
                    let scalar = Unicode.Scalar(value)
                {
                    out.unicodeScalars.append(scalar)
                } else {
                    out.append(contentsOf: text[index...semicolon.lowerBound])
                }
            }
            index = semicolon.upperBound
        }
        return out
    }

    // MARK: - Workbook

    /// The shared string table, where most cell text actually lives.
    ///
    /// An entry can be split across several `<t>` runs when parts of it are formatted
    /// differently, so runs are joined rather than the first taken — `Mª DEL CARMEN`
    /// arrives in two runs in the Spanish file.
    static func sharedStrings(_ xml: String?) -> [String] {
        guard let xml else { return [] }
        return elements("si", in: xml).map { entry in
            decodeEntities(elements("t", in: entry.body).map(\.body).joined())
        }
    }

    /// `B7` -> (row 7, column 1), the column zero-based.
    static func position(of reference: String) -> (row: Int, column: Int)? {
        var letters = ""
        var digits = ""
        for character in reference {
            if character.isUppercase && character.isLetter && digits.isEmpty {
                letters.append(character)
            } else if character.isNumber {
                digits.append(character)
            } else {
                return nil
            }
        }
        guard !letters.isEmpty, let row = Int(digits) else { return nil }
        var column = 0
        for character in letters.unicodeScalars {
            column = column * 26 + Int(character.value) - 64
        }
        return (row, column - 1)
    }

    /// Sheet name to XML part, resolved through the workbook and its relationships.
    static func sheetParts(workbook: String?, relationships: String?) -> [(String, String)] {
        var targets: [String: String] = [:]
        for relationship in elements("Relationship", in: relationships ?? "") {
            guard let id = attribute("Id", in: relationship.attributes),
                var target = attribute("Target", in: relationship.attributes)
            else { continue }
            if target.hasPrefix("/xl/") { target = String(target.dropFirst(4)) }
            else if target.hasPrefix("xl/") { target = String(target.dropFirst(3)) }
            else if target.hasPrefix("/") { target = String(target.dropFirst()) }
            targets[id] = target
        }

        var parts: [(String, String)] = []
        for sheet in elements("sheet", in: workbook ?? "") {
            guard let name = attribute("name", in: sheet.attributes),
                let id = attribute("r:id", in: sheet.attributes),
                let target = targets[id]
            else { continue }
            parts.append((decodeEntities(name), "xl/\(target)"))
        }
        return parts
    }

    /// Reads a workbook into sheet name -> rows, each row an array of cell strings.
    ///
    /// Values come back as strings whatever the cell held, because every caller is about to
    /// parse them itself and a number that arrived as `1063756` is not more trustworthy for
    /// having passed through a float.
    public static func readWorkbook(at path: URL) throws -> [String: [[String]]] {
        let shared = sharedStrings(try member("xl/sharedStrings.xml", in: path))
        let parts = sheetParts(
            workbook: try member("xl/workbook.xml", in: path),
            relationships: try member("xl/_rels/workbook.xml.rels", in: path))
        guard !parts.isEmpty else {
            throw Failure.missingMember("xl/workbook.xml", in: path.lastPathComponent)
        }

        var sheets: [String: [[String]]] = [:]
        for (name, part) in parts {
            guard let xml = try member(part, in: path) else { continue }
            var rows: [[String]] = []

            // selfClosing: false — see `elements`. A spacer row has no closing tag and is
            // not a row.
            for row in elements("row", in: xml, selfClosing: false) {
                var cells: [Int: String] = [:]
                var appended: [String] = []
                var highest = -1

                for cell in elements("c", in: row.body) {
                    let reference = attribute("r", in: cell.attributes)
                    let spot = reference.flatMap { position(of: $0) }
                    let type = attribute("t", in: cell.attributes)

                    var value = ""
                    if !cell.body.isEmpty {
                        if type == "inlineStr" {
                            value = decodeEntities(
                                elements("t", in: cell.body).map(\.body).joined())
                        } else if let raw = elements("v", in: cell.body).first?.body {
                            value =
                                type == "s"
                                ? (Int(raw).flatMap { $0 < shared.count ? shared[$0] : nil } ?? "")
                                : decodeEntities(raw)
                        }
                    }

                    // Placed by its own column rather than appended: a sheet omits empty
                    // cells entirely, and appending would slide everything after a gap one
                    // column to the left.
                    if let spot {
                        cells[spot.column] = value
                        highest = max(highest, spot.column)
                    } else {
                        appended.append(value)
                    }
                }

                let assembled = highest >= 0 ? (0...highest).map { cells[$0] ?? "" } : []
                rows.append(assembled + appended)
            }
            sheets[name] = rows
        }
        return sheets
    }
}
